# pandas 데이터 전처리 - apply, lambda, map 함수

## 데이터프레임으로 csv 파일 로딩

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('datas/korea-star.csv')
df

이름     그룹      소속사  성별        생년월일      키 혈액형      인기지수
0    아이린   레드벨벳       SM  여자  1991-03-29  160.0   A   8256324
1    차은우   아스트로     판타지오  남자  1997-03-30  183.0   B   3506027
2      뷔  방탄소년단      빅히트  남자  1995-12-30  178.0  AB   8073501
3     민현    NaN     플레디스  남자  1995-08-09  182.3   O   4989792
4     소연    아이들       큐브  여자  1998-08-26    NaN   B   4668615
5      진  방탄소년단      빅히트  남자  1992-12-04  179.2   O   4570308
6     윤아   소녀시대       SM  여자  1989-03-09    NaN   A   3918661
7     백호   뉴이스트     플레디스  남자  1995-07-21  175.0  AB   3301654
8     JR   뉴이스트     플레디스  남자  1995-06-08  176.0   O   3274137
9     슈가  방탄소년단      빅히트  남자  1993-03-09  174.0   O   4925442
10  강다니엘    NaN      커넥트  남자  1996-12-10  180.0   A   8273745
11    정국  방탄소년단      빅히트  남자  1997-09-01  178.0   A   7208335
12    화사    마마무      RBW  여자  1995-07-23  162.1   A   7650928
13   하성운     핫샷  스타크루이엔티  남자  1994-03-22  167.1   A   4036489
14    슬기   레드벨벳       SM  여자  1994-02-10  161.0   A   5021452
15    지민  방탄소년단      빅히트  남자  1995-10-13  173.6   A  10523260

# 1. apply()함수 사용
- Series나 DataFrame에 좀 더 구체적인 로직을 적용하고 싶은 경우 활용

- apply를 적용하기 위해서는 함수가 먼저 정의되어야함.

- apply는 정의한 로직 함수를 인자로 넘겨줌

## 1-1 loc를 활용한 범주값 숫자로 변경하기

In [3]:
df.loc[df['성별'] == '남자', '성별'] = 1
df.loc[df['성별'] == '여자', '성별'] = 0

In [4]:
df

이름     그룹      소속사 성별        생년월일      키 혈액형      인기지수
0    아이린   레드벨벳       SM  0  1991-03-29  160.0   A   8256324
1    차은우   아스트로     판타지오  1  1997-03-30  183.0   B   3506027
2      뷔  방탄소년단      빅히트  1  1995-12-30  178.0  AB   8073501
3     민현    NaN     플레디스  1  1995-08-09  182.3   O   4989792
4     소연    아이들       큐브  0  1998-08-26    NaN   B   4668615
5      진  방탄소년단      빅히트  1  1992-12-04  179.2   O   4570308
6     윤아   소녀시대       SM  0  1989-03-09    NaN   A   3918661
7     백호   뉴이스트     플레디스  1  1995-07-21  175.0  AB   3301654
8     JR   뉴이스트     플레디스  1  1995-06-08  176.0   O   3274137
9     슈가  방탄소년단      빅히트  1  1993-03-09  174.0   O   4925442
10  강다니엘    NaN      커넥트  1  1996-12-10  180.0   A   8273745
11    정국  방탄소년단      빅히트  1  1997-09-01  178.0   A   7208335
12    화사    마마무      RBW  0  1995-07-23  162.1   A   7650928
13   하성운     핫샷  스타크루이엔티  1  1994-03-22  167.1   A   4036489
14    슬기   레드벨벳       SM  0  1994-02-10  161.0   A   5021452
15    지민  방탄소년단      빅히트  1  1995-10-13  173.6   A  10523260

## 1-2. apply () - 정의한 함수 DataFrame에 일괄적용
- 남자: 1
- 여자: 0
- 그 외: -1


**[주의]**

반드시 return 값이 존재 해야함  
apply : pandas dataFrame 함수 == python map

In [15]:
df = pd.read_csv('datas/korea-star.csv')
df

이름     그룹      소속사  성별        생년월일      키 혈액형      인기지수
0    아이린   레드벨벳       SM  여자  1991-03-29  160.0   A   8256324
1    차은우   아스트로     판타지오  남자  1997-03-30  183.0   B   3506027
2      뷔  방탄소년단      빅히트  남자  1995-12-30  178.0  AB   8073501
3     민현    NaN     플레디스  남자  1995-08-09  182.3   O   4989792
4     소연    아이들       큐브  여자  1998-08-26    NaN   B   4668615
5      진  방탄소년단      빅히트  남자  1992-12-04  179.2   O   4570308
6     윤아   소녀시대       SM  여자  1989-03-09    NaN   A   3918661
7     백호   뉴이스트     플레디스  남자  1995-07-21  175.0  AB   3301654
8     JR   뉴이스트     플레디스  남자  1995-06-08  176.0   O   3274137
9     슈가  방탄소년단      빅히트  남자  1993-03-09  174.0   O   4925442
10  강다니엘    NaN      커넥트  남자  1996-12-10  180.0   A   8273745
11    정국  방탄소년단      빅히트  남자  1997-09-01  178.0   A   7208335
12    화사    마마무      RBW  여자  1995-07-23  162.1   A   7650928
13   하성운     핫샷  스타크루이엔티  남자  1994-03-22  167.1   A   4036489
14    슬기   레드벨벳       SM  여자  1994-02-10  161.0   A   5021452
15    지민  방탄소년단      빅히트  남자  1995-10-13  173.6   A  10523260

In [11]:
# 먼저 apply에 활용할 함수를 정의합니다.
def gender(x):
    if x == '남자':
        return 1
    elif x == '여자':
        return 0
    else:
        return -1

In [10]:
# python map() 함수 사용
def a_func(x):
    return x * 2

a = [1, 2, 3, 4]

result = map(a_func, a)
# for test in result:
#     print(test) : map 도 덩어리 : 반복문 사용
print(result)
print(type(result))
print(list(result))

<class 'map'>
[2, 4, 6, 8]


In [16]:
# df['성별'] = df['성별'].apply(gender)
df['성별_NEW'] = df['성별'].apply(gender)

In [17]:
df.head()

이름     그룹   소속사  성별        생년월일      키 혈액형     인기지수  성별_NEW
0  아이린   레드벨벳    SM  여자  1991-03-29  160.0   A  8256324       0
1  차은우   아스트로  판타지오  남자  1997-03-30  183.0   B  3506027       1
2    뷔  방탄소년단   빅히트  남자  1995-12-30  178.0  AB  8073501       1
3   민현    NaN  플레디스  남자  1995-08-09  182.3   O  4989792       1
4   소연    아이들    큐브  여자  1998-08-26    NaN   B  4668615       0

## [문제] apply 함수 활용 및 컬럼 추가
- (목표) cm당 인기지수 구하기
- (인기지수/ 키)를 구하여 '인기지수/키' 컬럼 추가

ex
- 키: 174, 인기지수: 100000
- 계산 식: 100000 / 174

In [126]:
df = pd.read_csv('datas/korea-star.csv')

### 1. apply에 적용할할 함수 정의

In [151]:
def cm_to_brand1(x):
    # print(x["키"])
    result = x["인기지수"]/x["키"]
    # print(result)
    return round(result, 2)

In [116]:
def cm_to_brand(x):
    result = x["인기지수"] / x["키"]
    return result

### 2. df.apply(호출함수) 실행

In [152]:
df["키/인기지수"] = df[["인기지수", "키"]].apply(cm_to_brand1, axis=1)

In [117]:
df["키/인기지수"] = df.apply(cm_to_brand, axis=1)

In [148]:
df

이름     그룹      소속사  성별        생년월일      키 혈액형      인기지수    키/인기지수
0    아이린   레드벨벳       SM  여자  1991-03-29  160.0   A   8256324  51602.02
1    차은우   아스트로     판타지오  남자  1997-03-30  183.0   B   3506027  19158.62
2      뷔  방탄소년단      빅히트  남자  1995-12-30  178.0  AB   8073501  45356.75
3     민현    NaN     플레디스  남자  1995-08-09  182.3   O   4989792  27371.32
4     소연    아이들       큐브  여자  1998-08-26    NaN   B   4668615       NaN
5      진  방탄소년단      빅히트  남자  1992-12-04  179.2   O   4570308  25503.95
6     윤아   소녀시대       SM  여자  1989-03-09    NaN   A   3918661       NaN
7     백호   뉴이스트     플레디스  남자  1995-07-21  175.0  AB   3301654  18866.59
8     JR   뉴이스트     플레디스  남자  1995-06-08  176.0   O   3274137  18603.05
9     슈가  방탄소년단      빅히트  남자  1993-03-09  174.0   O   4925442  28307.14
10  강다니엘    NaN      커넥트  남자  1996-12-10  180.0   A   8273745  45965.25
11    정국  방탄소년단      빅히트  남자  1997-09-01  178.0   A   7208335  40496.26
12    화사    마마무      RBW  여자  1995-07-23  162.1   A   7650928  47198.82
13   하성운     핫샷  스타크루이엔티  남자  1994-03-22  167.1   A   4036489  24156.13
14    슬기   레드벨벳       SM  여자  1994-02-10  161.0   A   5021452  31189.14
15    지민  방탄소년단      빅히트  남자  1995-10-13  173.6   A  10523260  60617.86

## 1-3. lambda 함수의 적용
- lambda 함수 : 한줄 함수
- 식별자(변수) = lambda 매개변수1, 매개변수2 : 수행문
- f = lambda x, y : x + y

In [153]:
# 실행결과 식별 = f(5, 6)

f = lambda x, y : x+y
f(5, 6)

11

In [173]:
def lambda_ex(num1, num2):
    return num1 * num2
lambda_ex(f(6, ), 2)

0

In [191]:
def lambda_ex(num1, num2):
    return num1 * num2

test = [1]
test1 = [2] 

lambda_ex(int(list(map(lambda x, y : x+y, test, test1))[0]), 2)

6

## lambda 함수 직접 적용하기

In [170]:
c = [1, 2, 3, 4]
map(lambda x: x*2, c)

In [171]:
list(map(lambda x: x*2, c))

[2, 4, 6, 8]

In [161]:
df["키/2"] = df["키"].apply(lambda x: x/2)

In [163]:
df.head(3)

이름     그룹   소속사  성별        생년월일      키 혈액형     인기지수    키/인기지수   키/2
0  아이린   레드벨벳    SM  여자  1991-03-29  160.0   A  8256324  51602.02  80.0
1  차은우   아스트로  판타지오  남자  1997-03-30  183.0   B  3506027  19158.62  91.5
2    뷔  방탄소년단   빅히트  남자  1995-12-30  178.0  AB  8073501  45356.75  89.0

In [165]:
f = lambda x: 1 if x == "남자" else 0

In [166]:
df["성별"].apply(f)

0     0
1     1
2     1
3     1
4     0
5     1
6     0
7     1
8     1
9     1
10    1
11    1
12    0
13    1
14    0
15    1
Name: 성별, dtype: int64

In [167]:
df["키"].apply(lambda x: x**2)

0     25600.00
1     33489.00
2     31684.00
3     33233.29
4          NaN
5     32112.64
6          NaN
7     30625.00
8     30976.00
9     30276.00
10    32400.00
11    31684.00
12    26276.41
13    27922.41
14    25921.00
15    30136.96
Name: 키, dtype: float64

### 정리 : apply() 함수식을 만들어서 적용해 줌.
- 복잡한 조건식은 함수 정의로
- 간단한 계산식은 lambda로 적용

# 3. map() - 값을 매핑!

In [211]:
df = pd.read_csv('datas/korea-star.csv')
df.head()

이름     그룹   소속사  성별        생년월일      키 혈액형     인기지수
0  아이린   레드벨벳    SM  여자  1991-03-29  160.0   A  8256324
1  차은우   아스트로  판타지오  남자  1997-03-30  183.0   B  3506027
2    뷔  방탄소년단   빅히트  남자  1995-12-30  178.0  AB  8073501
3   민현    NaN  플레디스  남자  1995-08-09  182.3   O  4989792
4   소연    아이들    큐브  여자  1998-08-26    NaN   B  4668615

#### df.map(mapping) 상용 방법
- mapping: dictionary
<pre>
{
    원래값1:변경값1,
    원래값2:변경값2,
    ...
    원래값n:변경값n
}
</pre>

In [202]:
print(df["성별"].value_counts()) # 정보 확인
print(df["성별"].unique())

성별
남자    11
여자     5
Name: count, dtype: int64
['여자' '남자']


In [217]:
mapping = {
    "여자" : 0,
    "남자" : 1
}

In [220]:
df["성별"] = df["성별"].map(mapping)

In [206]:
df.head()

이름     그룹   소속사  성별        생년월일      키 혈액형     인기지수
0  아이린   레드벨벳    SM   0  1991-03-29  160.0   A  8256324
1  차은우   아스트로  판타지오   1  1997-03-30  183.0   B  3506027
2    뷔  방탄소년단   빅히트   1  1995-12-30  178.0  AB  8073501
3   민현    NaN  플레디스   1  1995-08-09  182.3   O  4989792
4   소연    아이들    큐브   0  1998-08-26    NaN   B  4668615

In [221]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16 entries, 0 to 15
Data columns (total 8 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   이름      16 non-null     object 
 1   그룹      14 non-null     object 
 2   소속사     16 non-null     object 
 3   성별      0 non-null      float64
 4   생년월일    16 non-null     object 
 5   키       14 non-null     float64
 6   혈액형     16 non-null     object 
 7   인기지수    16 non-null     int64  
dtypes: float64(2), int64(1), object(5)
memory usage: 1.1+ KB


In [210]:
# mapping 원래값 type을 맞춰줘야 함
mapping = {
    "0": "female",
    "1": "male"
}

In [209]:
df["성별"] = df["성별"].map(mapping)
df.head()

이름     그룹   소속사   성별        생년월일      키 혈액형     인기지수
0  아이린   레드벨벳    SM  NaN  1991-03-29  160.0   A  8256324
1  차은우   아스트로  판타지오  NaN  1997-03-30  183.0   B  3506027
2    뷔  방탄소년단   빅히트  NaN  1995-12-30  178.0  AB  8073501
3   민현    NaN  플레디스  NaN  1995-08-09  182.3   O  4989792
4   소연    아이들    큐브  NaN  1998-08-26    NaN   B  4668615

In [214]:
mapping = {
    0: "female",
    1: "male"
}

In [215]:
df["성별"] = df["성별"].map(mapping)
df.head()

이름     그룹   소속사      성별        생년월일      키 혈액형     인기지수
0  아이린   레드벨벳    SM  female  1991-03-29  160.0   A  8256324
1  차은우   아스트로  판타지오    male  1997-03-30  183.0   B  3506027
2    뷔  방탄소년단   빅히트    male  1995-12-30  178.0  AB  8073501
3   민현    NaN  플레디스    male  1995-08-09  182.3   O  4989792
4   소연    아이들    큐브  female  1998-08-26    NaN   B  4668615

In [222]:
df["성별"].dtypes

dtype('float64')